component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/EventHubClient.java

```diff
@@ -32,74 +45,218 @@
+                            //For any other exception
+                            log.error("Unknown error occurred while publishing EventDataBatch. Producer client will "
+                                              + "be re-initialized. Events may be lost in the process.");
+                            this.clientStatus = ClientStatus.RETRYING;
+                            readWriteLock.writeLock().unlock();
+                            createProducerWithRetry(authEndpoint, authToken, retryOptions, true);
+                        }
+                    } finally {
+                        try {
+                            readWriteLock.writeLock().unlock();
+                        } catch (IllegalMonitorStateException e) {
+                            //ignore
+                        }
                     }
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                } finally {
-                    sendSemaphore.release();
+                }
+                if (isAdded) {
+                    log.debug("Adding event: " + event.replaceAll("[\r\n]", ""));
+                } else {
+                    log.debug("Failed to add event: " + event.replaceAll("[\r\n]", ""));
+                }
+            } finally {
+                try {
+                    readWriteLock.readLock().unlock();
+                } catch (IllegalMonitorStateException e) {
+                    //ignore
                 }
             }
+        } else {
+            try {
+                threadBarrier.lock();
+                log.debug(Thread.currentThread().getName().replaceAll("[\r\n]", "") + " will be parked as EventHub "
+                                  + "Client is inactive.");
+                waitCondition.await();
+                log.debug(Thread.currentThread().getName().replaceAll("[\r\n]", "") + " will be resumes as EventHub "
+                                  + "Client is active.");
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } finally {
+                threadBarrier.unlock();
+            }
+            sendEvent(event);
+        }
+    }
+
+    private boolean isAuthenticationFailure(AmqpException exception) {
+        AmqpErrorCondition condition = exception.getErrorCondition();
+        return (condition == AmqpErrorCondition.UNAUTHORIZED_ACCESS ||
+                condition == AmqpErrorCondition.PUBLISHER_REVOKED_ERROR);
+    }
 
-            if (isAdded) {
-                log.debug("Adding event: " + event.replaceAll("[\r\n]", ""));
-            } else {
-                log.debug("Failed to add event: " + event.replaceAll("[\r\n]", ""));
+    private EventDataBatch createBatchWithRetry() {
+        try {
+            EventDataBatch batch = producer.createBatch();
+            eventBatchRetryCounter.reset();
+            return batch;
+        } catch (IllegalStateException e) {
```

Comment:
From where do we get this RuntimeException?

Reply:
producer.createBatch() will throw this when batch creation fails